packages/bs-solana/src/services/blockchain-data/TatumRpcBDSSolana.ts

```diff
@@ -0,0 +1,436 @@
+
+    return {
+      type: 'token',
+      amount,
+      from: info.source,
+      to: info.destination,
+      contractHash: BSSolanaConstants.NATIVE_TOKEN.hash,
+      token: BSSolanaConstants.NATIVE_TOKEN,
+    }
+  }
+
+  async #parseInstruction(
+    instruction: solanaSDK.ParsedInstruction,
+    allInstructions: solanaSDK.ParsedInstruction[]
+  ): Promise<TransactionTransferAsset | TransactionTransferNft> {
+    if (!instruction.parsed.type || !instruction.parsed.info) {
+      throw new Error('Unsupported instruction format')
+    }
+
+    const programId = instruction.programId.toString()
+    const method = instruction.parsed.type
+
+    const func = this.#functionByProgramIdAndMethod[programId]?.[method]
+    if (func) {
+      return await func(instruction, allInstructions)
+    }
+
+    throw new Error('Unsupported instruction')
+  }
+
+  async getTransaction(txid: string): Promise<TransactionResponse> {
+    const transaction = await this.#connection.getParsedTransaction(txid, {
+      maxSupportedTransactionVersion: 0,
+    })
+
+    if (!transaction) throw new Error('Transaction not found')
+
+    if (!transaction.blockTime) throw new Error('Block time not found')
+    if (!transaction.meta) throw new Error('Transaction meta not found')
+
+    const transfers: (TransactionTransferAsset | TransactionTransferNft)[] = []
+
+    const allInstructions = [
+      ...transaction.transaction.message.instructions,
+      ...(transaction.meta.innerInstructions?.flatMap(item => item.instructions) ?? []),
+    ].filter((item): item is solanaSDK.ParsedInstruction => 'parsed' in item)
+
+    for (const instruction of allInstructions) {
+      try {
+        const transfer = await this.#parseInstruction(instruction, allInstructions)
+        if (transfer) transfers.push(transfer)
+      } catch {
+        /* empty */
+      }
+    }
+
+    return {
+      block: transaction.slot,
+      hash: transaction.transaction.signatures[0],
+      notifications: [],
+      time: new Date(transaction.blockTime).getTime() / 1000,
+      transfers,
+      fee: formatTokenNumber(transaction.meta.fee, this.#feeToken.decimals),
+    }
+  }
+
+  async getTransactionsByAddress(params: TransactionsByAddressParams): Promise<TransactionsByAddressResponse> {
+    const publicKey = new solanaSDK.PublicKey(params.address)
+
+    const signaturesResponse = await this.#connection.getSignaturesForAddress(publicKey, {
+      limit: 15,
+      before: params.nextPageParams,
+    })
+
+    const transactions: TransactionResponse[] = []
+
+    const promises = signaturesResponse.map(async signatureResponse => {
+      const transaction = await this.getTransaction(signatureResponse.signature)
+      transactions.push(transaction)
+    })
+
+    await Promise.allSettled(promises)
+
+    return {
+      transactions,
+      nextPageParams: signaturesResponse.length === 15 ? signaturesResponse.at(-1)?.signature : undefined,
```

Comment:
I don't really like this validation, because the last page could have 15 items and the next 0 items.

Reply:
Do you know a alternative ?
I can't find any another way to do that

Reply:
Does the signature always exist?